components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/client/CallBackValidator.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.mgt.endpoint.client;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.base.MultitenantConstants;
+import org.wso2.carbon.identity.application.common.model.IdentityProvider;
+import org.wso2.carbon.identity.application.common.model.IdentityProviderProperty;
+import org.wso2.carbon.identity.mgt.endpoint.IdentityManagementEndpointConstants;
+import org.wso2.carbon.idp.mgt.IdentityProviderManagementException;
+import org.wso2.carbon.idp.mgt.IdentityProviderManager;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+
+/**
+ * Validates the call back URL passed in the recovery request.
+ */
+public class CallBackValidator {
+
+    private static final Log log = LogFactory.getLog(CallBackValidator.class);
+
+    /**
+     * This method is to validate the callback URL in the request with the configured one.
+     *
+     * @param callbackURL  passed in the request
+     * @param tenantDomain of the user
+     * @return the status of the validation
+     * @throws IdentityRecoveryException
+     */
+    public boolean isValidCallbackURL(String callbackURL, String tenantDomain, boolean isUserPortalURL)
+            throws IdentityRecoveryException {
+
+        if (isUserPortalURL) {
+            if (log.isDebugEnabled()) {
+                log.debug("Callback URL is equal to the user portal URL: " + isUserPortalURL);
```

Comment:
this debug will be logged only if isUserPortalURL is true. Shall we add a debug log to trace the false scenario?

Reply:
user portal URL will be assigned to the callback URL by the recovery endpoint only when callback URL is null. isURLPortalURL will be passed to the server only to notify that the callback URL is equal to userPortalURL to avoid callback validation since this will be handled internally by the server. Else case will be logged at https://github.com/wso2/carbon-identity-framework/pull/2075/files/799f4304c4a890d0e0d0e880e9325f54640ecf62#diff-a0ef0b46165988efc2915a3ae61f2b70R103.